### PR TITLE
Limit the CSP header size to avoid server errors

### PR DESCRIPTION
**Pull request by Toflar**
With the latest PRs merged in the 5.3 branch, the probability of generating a large CSP header increased but it was also possible before: the CSP header can get pretty long. If you e.g. have a 50 bytes long inline style, a sha384 hash base64 encoded can get about 130 bytes long. (`echo strlen('sha384-' . base64_encode(hash('sha384', random_bytes(50))));`).

I found various different sources for the maximum header size of nginx, Apache and Co. but it ranges from 4kb to 16kb. If it is 4kb, you'll end up having a max of about 30 of those hashes and you will end up crashing the webserver for a header too long. 

I think, it would be great if Contao could limit this. It should be better to remove hashes and have CSP violations than having a server crash. Or in other words: better to not have a `text-decoration: underline` than a 500 Server Error.

So here's some proposal reducing the CSP header length automatically, starting with `style-src` hashes and then `script-src` hashes until the header size is okay.

### CODE REVIEW

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -875,6 +875,10 @@ static function (array $allowedProperties): array {
                         )
                     ->end()
                 ->end()
+                ->integerNode('max_header_size')
+                    ->info('The CSP header can get pretty long when adding lots of automatically calculated hashes. Contao will automatically remove signatures in order to prevent the CSP header to exceed the configured limit. This will affect the functionality of affected URLs but ensure you do not run into 500 server errors.')
+                    ->defaultValue(4096) // Nginx default maximum header length
```
Comment (Toflar) [inline core-bundle/src/DependencyInjection/Configuration.php]: Maybe this could be 8kb. See http://nginx.org/en/docs/http/ngx_http_core_module.html

File: core-bundle/src/Routing/ResponseContext/Csp/CspHandler.php [hunk 2]
```diff
@@ -158,4 +162,66 @@ public function applyHeaders(Response $response, Request|null $request = null):
...
+        if (!isset($this->signatures[$source])) {
+            return;
+        }
+
+        // Unset the ones added last first (in case of style-src, that would mean that the top inline styles would still
+        // work while towards the footer they might not work anymore)
+        do {
+            array_pop($this->signatures[$source]);
+        } while (\strlen($this->buildHeaderValue()) > $this->maxHeaderLength && [] !== $this->signatures[$source]);
+    }
```
Comment (fritzmg) [inline core-bundle/src/Routing/ResponseContext/Csp/CspHandler.php]: Should we add some logging here to notify that the header needed to be reduced?

File: core-bundle/src/Twig/Runtime/CspRuntime.php [hunk 3]
```diff
@@ -42,6 +42,10 @@ public function inlineStyles(string $htmlFragment): string
             return $htmlFragment;
         }
 
+        if (!$styles = $this->wysiwygProcessor->extractStyles($htmlFragment)) {
+            return $htmlFragment;
+        }
+
```
Comment (ausi) [inline core-bundle/src/Twig/Runtime/CspRuntime.php]:
```suggestion
```

Looks like a failed rebase/merge as this code block is the same as the one above.

File: core-bundle/config/services.yaml [hunk 4]
```diff
@@ -806,9 +806,18 @@ services:
             - '@contao.string.html_decoder'
             - '@request_stack'
             - '@contao.insert_tag.parser'
-            - '@contao.csp.parser'
+            - '@contao.routing.response_context.csp.handler_factory'
             - '@router'
 
+    contao.routing.response_context.csp.handler_factory:
+        class: Contao\CoreBundle\Routing\ResponseContext\Csp\CspHandlerFactory
```
Comment (ausi) [inline core-bundle/config/services.yaml]: Looks like this `CspHandlerFactory` class is missing.

File: core-bundle/src/Routing/ResponseContext/Csp/CspHandler.php (hunk 2 above)
Comment (ausi) [reply core-bundle/src/Routing/ResponseContext/Csp/CspHandler.php]: I think we should add information like this to the existing `$this->logger?->critical(sprintf('Allowed CSP header size of %d bytes exhausted…`

Comment (Toflar) [reply core-bundle/src/Routing/ResponseContext/Csp/CspHandler.php]: Like this? 57b9d9c53cf7059c5554430404244fb372230e0c

**Comment by Toflar** [comment]
Should be ready. The only thing here missing is a decision for the default length. Looks like nginx is 8kb, as is Apache. So maybe we could increase the default to 8192? /cc @leofeyer

File: core-bundle/src/DependencyInjection/Configuration.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/DependencyInjection/Configuration.php]: Yes, that should be fine. 👍

**Comment by ausi** [comment]
Most information about header size I could find online talks about the size of the whole HTTP header block containing all headers. Do we maybe need to take the size of the other headers in the `$response` into account when reducing the size?

> The only thing here missing is a decision for the default length. Looks like nginx is 8kb, as is Apache. So maybe we could increase the default to 8192?

Nginx writes _“By default, the buffer size is equal to one memory page. This is either 4K or 8K, depending on a platform.”_
I have no idea what a common memory page size is on most webservers but searching the internet 4K seems to be quite common.

**Comment by fritzmg** [comment]
Yes, the maximum header size relates to the sum of all headers, not just one single header.

To be honest, I am not sure anymore we should or can implement this. You'll have the same issue if you have too many / too large cookies or may be even too many cache tags.

**Comment by Toflar** [comment]
> Yes, the maximum header size relates to the sum of all headers, not just one single header.
> 
> To be honest, I am not sure anymore we should or can implement this. You'll have the same issue if you have too many / too large cookies or may be even too many cache tags.

Then we must implement this PR even more so! I don't think considering the other headers makes any sense. We should just limit the CSP header to a fixed amount of bytes for maximum compatibility. This is just a safety net anyway. Most users won't hit it because they will not have 30 hashes. Remember, if you have multiple tinyMCE values on the same page, all using `text-decoration: underline` only, you'll end up having only one single hash. Only if you use multiple `style` different style combinations you might run into this issue. And I think this PR greatly enhances usability. It will log and tell you that the CSP header was too long and what you should do. If you run into a 502 or 500, you'll have to search the server logs and all you'll probably find is that the total header content was too long, not giving you any hints about the CSP header being the culprit or not at all.

**Comment by ausi** [comment]
> We should just limit the CSP header to a fixed amount of bytes for maximum compatibility.

Agree. How many bytes do we want to allow by default? 3072 or 2048?

**Comment by Toflar** [comment]
> Agree. How many bytes do we want to allow by default? 3072 or 2048?

I wrote a quick calculation script with a realistic base CSP header and the options to adjust the SHA algorithm and max header length to see how many hashes would fit: https://3v4l.org/YPirH

I think we should probably switch to SHA256 anyway (secure enough, speed is probably neglectable and should be in cache for public pages anyway) and 3KB. This would allow a max of 29 hashes which I think is easily enough for most use cases.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php [hunk 5]
```diff
@@ -144,15 +144,14 @@ private function addCspHandler(ResponseContext $context, PageModel $pageModel):
 
             try {
                 $reportUri = $this->urlGenerator->generate(CspReporterController::class, ['page' => $pageModel->id], UrlGeneratorInterface::ABSOLUTE_URL);
-                $directives->setDirective('report-uri', $reportUri);
+                $cspHandler->getDirectives()->setDirective('report-uri', $reportUri);
```
Comment (leofeyer) [inline core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php:147]: Why did you move this into the `try` block?

Comment (Toflar) [reply core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php:147]: I did not move anything in the try block? It is where it was before, only the `$directives` does not exist anymore.

Comment (leofeyer) [reply core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php:147]: True, my bad.

**Review by leofeyer — CHANGES_REQUESTED** [review]
It seems the changes have caused a problem in the `applyHeaders()` method:

**Comment by Toflar** [comment]
Ah that's a leftover, fixed in https://github.com/contao/contao/pull/6761/commits/cd6f3f755331b4d8d9076932e9ecab034f431ff5

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -875,6 +875,10 @@ static function (array $allowedProperties): array {
                         )
                     ->end()
                 ->end()
+                ->integerNode('max_header_size')
+                    ->info('The CSP header can get pretty long when adding lots of automatically calculated hashes. Contao will automatically remove signatures in order to prevent the CSP header to exceed the configured limit. This will affect the functionality of affected URLs but ensure you do not run into 500 server errors.')
+                    ->defaultValue(3072) // Most browsers have a limit of 8-16kb for **all** headers in total
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Configuration.php]: This describes the technical background of the pull request rather than the `max_header_size` parameter, which is not very helpful for someone who wants to adjust the parameter and maybe needs some pointers, is it?

```suggestion
                    ->info('Do not increase this value beyond the allowed response headers size of your web server, as this will result in a 500 server error.')
                    ->defaultValue(3072)
```

Comment (Toflar) [reply core-bundle/src/DependencyInjection/Configuration.php]: Are you sure it would end in a 400? Sounds like it should be a 500 as it's not the client's fault?

Comment (leofeyer) [reply core-bundle/src/DependencyInjection/Configuration.php]: I haven‘t tried it, but it‘s what the Nginx docs say:

> A request line cannot exceed the size of one buffer, or the 414 (Request-URI Too Large) error is returned to the client. A request header field cannot exceed the size of one buffer as well, or the 400 (Bad Request) error is returned to the client.

https://nginx.org/en/docs/http/ngx_http_core_module.html#large_client_header_buffers

Comment (Toflar) [reply core-bundle/src/DependencyInjection/Configuration.php]: Those are request headers. We are talking about response headers :) Pretty sure it would be something else in our case. Would you mind trying it out on one of your nginx setups? 😇

Comment (leofeyer) [reply core-bundle/src/DependencyInjection/Configuration.php]: I have updated my suggestion accordingly.

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -875,6 +875,10 @@ static function (array $allowedProperties): array {
                         )
                     ->end()
                 ->end()
+                ->integerNode('max_header_size')
+                    ->info('Do not increase this value beyond the allowed response headers size of your web server, as this will result in a 500 server error.')
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Configuration.php]:
```suggestion
                    ->info('Do not increase this value beyond the allowed response header size of your web server, as this will result in a 500 server error.')
```

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.
